Multiple assemblies use the name "maven". Please assign each assembly a unique name. - using only single assembly

```
[DEBUG] Configuring mojo 'io.fabric8:docker-maven-plugin:0.40.0-SNAPSHOT:build' with basic configurator -->
[DEBUG]   (f) buildArchiveOnly = false
[DEBUG]   (f) execution = io.fabric8:docker-maven-plugin:0.40.0-SNAPSHOT:build {execution: default-cli}
[DEBUG]   (s) name = test
[DEBUG]   (f) from = java:8
[DEBUG]   (f) descriptor = docker-assembly.xml
[DEBUG]   (f) assembly = io.fabric8.maven.docker.config.AssemblyConfiguration@7fef0b40
[DEBUG]   (f) build = io.fabric8.maven.docker.config.BuildImageConfiguration@5f395ce1
[DEBUG]   (f) images = [ImageConfiguration {name='test', alias='null'}]
[DEBUG]   (f) jib = false
[DEBUG]   (f) jibImageFormat = docker
[DEBUG]   (f) keepContainer = false
[DEBUG]   (f) logStdout = false
[DEBUG]   (f) maxConnections = 100
[DEBUG]   (f) outputDirectory = target/docker
[DEBUG]   (f) packaging = pom
[DEBUG]   (f) project = MavenProject: test:test:1.0.0-SNAPSHOT @ pom-bug.xml
[DEBUG]   (f) reactorProjects = [MavenProject: test:test:1.0.0-SNAPSHOT @ pom-bug.xml]
[DEBUG]   (f) removeVolumes = false
[DEBUG]   (f) session = org.apache.maven.execution.MavenSession@b30a50d
[DEBUG]   (f) settings = org.apache.maven.execution.SettingsAdapter@764a3e5d
[DEBUG]   (f) skip = false
[DEBUG]   (f) skipBuild = false
[DEBUG]   (f) skipExtendedAuth = false
[DEBUG]   (f) skipMachine = false
[DEBUG]   (f) skipPom = false
[DEBUG]   (f) skipTag = false
[DEBUG]   (f) sourceDirectory = src/main/docker
[DEBUG]   (f) useColor = true
[DEBUG] -- end configuration --
[ERROR] DOCKER> Multiple assemblies use the name "maven". Please assign each assembly a unique name.
pom.xml
```

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>test</groupId>
    <artifactId>test</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <packaging>pom</packaging>

    <build>
        <plugins>
            <plugin>
                <groupId>io.fabric8</groupId>
                <artifactId>docker-maven-plugin</artifactId>
                <version>0.40.0-SNAPSHOT</version>
                <extensions>true</extensions>

                <configuration>
                    <images>
                        <image>
                            <name>test</name>
                            <build>
                                <from>java:8</from>
                                <assembly>
                                    <descriptor>docker-assembly.xml</descriptor>
                                </assembly>
                            </build>
                        </image>
                    </images>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

--- a/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
+++ b/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java
@@ -146,7 +146,7 @@ public class DockerAssemblyManager {
             throws MojoExecutionException {
 
         final BuildDirs buildDirs = createBuildDirs(imageName, params);
-        final List<AssemblyConfiguration> assemblyConfigurations = buildConfig.getAssemblyConfigurations();
+        final List<AssemblyConfiguration> assemblyConfigurations = buildConfig.getAllAssemblyConfigurations();
 
         final List<ArchiverCustomizer> archiveCustomizers = new ArrayList<>();
 
@@ -241,7 +241,7 @@ public class DockerAssemblyManager {
 
     // visible for testing
     void verifyGivenDockerfile(File dockerFile, BuildImageConfiguration buildConfig, FixedStringSearchInterpolator interpolator, Logger log) throws IOException {
-        List<AssemblyConfiguration> assemblyConfigs = buildConfig.getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblyConfigs = buildConfig.getAllAssemblyConfigurations();
         if (assemblyConfigs.isEmpty()) {
             return;
         }

--- a/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
+++ b/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
@@ -13,13 +13,15 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
+
 import javax.annotation.Nonnull;
+
+import org.apache.maven.plugins.annotations.Parameter;
 
 import io.fabric8.maven.docker.util.DeepCopy;
 import io.fabric8.maven.docker.util.EnvUtil;
 import io.fabric8.maven.docker.util.Logger;
 import io.fabric8.maven.docker.util.MojoParameters;
-import org.apache.maven.plugins.annotations.Parameter;
 
 /**
  * @author roland
@@ -295,16 +297,21 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     /**
-     * @deprecated Use {@link #getAssemblyConfigurations()} instead.
+     * @deprecated Use {@link #getAllAssemblyConfigurations()} instead.
      */
     @Deprecated
     public AssemblyConfiguration getAssemblyConfiguration() {
         return assembly;
     }
 
+
+    /**
+     * Use {@link #getAllAssemblyConfigurations()} unless you specifically want the configuration defined only by <code>assemblies</code>.
+     */
     @Nonnull
-    public List<AssemblyConfiguration> getAssemblyConfigurations() {
+    public List<AssemblyConfiguration> getAssembliesConfiguration() {
         final List<AssemblyConfiguration> assemblyConfigurations = new ArrayList<>();
+
         if (assemblies != null) {
             for (AssemblyConfiguration config : assemblies) {
                 if (config != null) {
@@ -312,9 +319,18 @@ public class BuildImageConfiguration implements Serializable {
                 }
             }
         }
+
+        return assemblyConfigurations;
+    }
+
+    @Nonnull
+    public List<AssemblyConfiguration> getAllAssemblyConfigurations() {
+        final List<AssemblyConfiguration> assemblyConfigurations = getAssembliesConfiguration();
+
         if (assembly != null) {
             assemblyConfigurations.add(assembly);
         }
+
         return assemblyConfigurations;
     }
 
@@ -762,7 +778,7 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     private void ensureUniqueAssemblyNames(Logger log) {
-        List<AssemblyConfiguration> assemblyConfigurations = getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblyConfigurations = getAllAssemblyConfigurations();
         Set<String> assemblyNames = new HashSet<>();
         for (AssemblyConfiguration config : assemblyConfigurations) {
             String assemblyName = config.getName();
@@ -822,7 +838,7 @@ public class BuildImageConfiguration implements Serializable {
         // TODO: Remove the following deprecated handling section
         if (dockerArchive == null) {
             Optional<String> deprecatedDockerFileDir =
-                    getAssemblyConfigurations().stream()
+                    getAllAssemblyConfigurations().stream()
                             .map(AssemblyConfiguration::getDockerFileDir)
                             .filter(Objects::nonNull)
                             .findFirst();

--- a/src/main/java/io/fabric8/maven/docker/config/handler/property/ConfigKey.java
+++ b/src/main/java/io/fabric8/maven/docker/config/handler/property/ConfigKey.java
@@ -26,6 +26,7 @@ public enum ConfigKey {
     ALIAS,
     ARGS(ValueCombinePolicy.Merge),
     ASSEMBLIES,
+    ASSEMBLY,
     ASSEMBLY_BASEDIR("assembly.baseDir"),
     ASSEMBLY_DESCRIPTOR("assembly.descriptor"),
     ASSEMBLY_DESCRIPTOR_REF("assembly.descriptorRef"),

--- a/src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java
+++ b/src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java
@@ -153,7 +153,7 @@ public class PropertyConfigHandler implements ExternalConfigHandler {
             .optimise(valueProvider.getBoolean(OPTIMISE, config.getOptimise()))
             .entryPoint(extractArguments(valueProvider, ENTRYPOINT, config.getEntryPoint()))
             .assembly(extractAssembly(config.getAssemblyConfiguration(), valueProvider))
-            .assemblies(extractAssemblies(config.getAssemblyConfigurations(), valueProvider))
+            .assemblies(extractAssemblies(config.getAssembliesConfiguration(), valueProvider))
             .env(CollectionUtils.mergeMaps(
                 valueProvider.getMap(ENV_BUILD, config.getEnv()),
                 valueProvider.getMap(ENV, Collections.emptyMap())
@@ -275,6 +275,12 @@ public class PropertyConfigHandler implements ExternalConfigHandler {
 
     @SuppressWarnings("deprecation")
     private AssemblyConfiguration extractAssembly(AssemblyConfiguration config, ValueProvider valueProvider) {
+        Map<String, String> assemblyProperties = valueProvider.getMap(ASSEMBLY, Collections.emptyMap());
+
+        if (assemblyProperties == null || assemblyProperties.isEmpty()) {
+            return config;
+        }
+
         if (config == null) {
             config = new AssemblyConfiguration();
         }

--- a/src/main/java/io/fabric8/maven/docker/service/ArchiveService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/ArchiveService.java
@@ -105,7 +105,7 @@ public class ArchiveService {
 
         String name = imageConfig.getName();
         try {
-            List<AssemblyConfiguration> assemblyConfigurations = imageConfig.getBuildConfiguration().getAssemblyConfigurations();
+            List<AssemblyConfiguration> assemblyConfigurations = imageConfig.getBuildConfiguration().getAllAssemblyConfigurations();
             AssemblyConfiguration assemblyConfig = assemblyConfigurations.stream().filter(a -> a.getName().equals(assemblyName)).findFirst().orElse(null);
 
             if (assemblyConfig == null) {

--- a/src/main/java/io/fabric8/maven/docker/service/BuildService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/BuildService.java
@@ -419,7 +419,7 @@ public class BuildService {
         String fromImage;
         fromImage = buildConfig.getFrom();
         if (fromImage == null) {
-            List<AssemblyConfiguration> assemblyConfig = buildConfig.getAssemblyConfigurations();
+            List<AssemblyConfiguration> assemblyConfig = buildConfig.getAllAssemblyConfigurations();
             if (assemblyConfig.isEmpty()) {
                 fromImage = DockerAssemblyManager.DEFAULT_DATA_BASE_IMAGE;
             }

--- a/src/main/java/io/fabric8/maven/docker/service/JibBuildService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/JibBuildService.java
@@ -59,7 +59,7 @@ public class JibBuildService {
 
             File dockerTarArchive = getAssemblyTarArchive(imageConfig, serviceHub, mojoParameters, log);
 
-            for (AssemblyConfiguration assemblyConfiguration : imageConfig.getBuildConfiguration().getAssemblyConfigurations()) {
+            for (AssemblyConfiguration assemblyConfiguration : imageConfig.getBuildConfiguration().getAllAssemblyConfigurations()) {
                 // TODO: Improve Assembly Manager so that the effective assemblyFileEntries computed can be properly shared
                 // the call to DockerAssemblyManager.getInstance().createDockerTarArchive should not be necessary,
                 // files should be added using the AssemblyFileEntry list. DockerAssemblyManager, should provide

--- a/src/main/java/io/fabric8/maven/docker/service/WatchService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/WatchService.java
@@ -78,7 +78,7 @@ public class WatchService {
                 ArrayList<String> tasks = new ArrayList<>();
 
                 if (imageConfig.getBuildConfiguration() != null) {
-                    for (AssemblyConfiguration assemblyConfiguration : imageConfig.getBuildConfiguration().getAssemblyConfigurations()) {
+                    for (AssemblyConfiguration assemblyConfiguration : imageConfig.getBuildConfiguration().getAllAssemblyConfigurations()) {
                         if (watcher.isCopy()) {
                             String containerBaseDir = assemblyConfiguration.getTargetDir();
                             schedule(executor, createCopyWatchTask(watcher, assemblyConfiguration.getName(), context.getMojoParameters(), containerBaseDir), interval);

--- a/src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java
+++ b/src/test/java/io/fabric8/maven/docker/assembly/DockerAssemblyManagerTest.java
@@ -31,7 +31,6 @@ import org.codehaus.plexus.components.io.resources.PlexusIoResource;
 import org.codehaus.plexus.interpolation.fixed.FixedStringSearchInterpolator;
 import org.codehaus.plexus.util.ReflectionUtils;
 import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.MethodOrderer;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.io.TempDir;
@@ -76,7 +75,7 @@ class DockerAssemblyManagerTest {
     @Test
     void testNoAssembly() {
         BuildImageConfiguration buildConfig = new BuildImageConfiguration();
-        List<AssemblyConfiguration> assemblyConfig = buildConfig.getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblyConfig = buildConfig.getAllAssemblyConfigurations();
 
         String content =
             assemblyManager.createDockerFileBuilder(
@@ -94,7 +93,7 @@ class DockerAssemblyManagerTest {
                 .build();
 
         DockerFileBuilder builder =
-            assemblyManager.createDockerFileBuilder(buildConfig, buildConfig.getAssemblyConfigurations());
+            assemblyManager.createDockerFileBuilder(buildConfig, buildConfig.getAllAssemblyConfigurations());
         String content = builder.content();
 
         Assertions.assertTrue(content.contains("SHELL [\"/bin/sh\",\"echo\",\"hello\"]"));
@@ -111,7 +110,7 @@ class DockerAssemblyManagerTest {
 
         BuildImageConfiguration buildConfig = createBuildConfig();
 
-        assemblyManager.getAssemblyFiles("testImage", buildConfig.getAssemblyConfigurations().get(0), mojoParams, new AnsiLogger(new SystemStreamLog(), true, "build"));
+        assemblyManager.getAssemblyFiles("testImage", buildConfig.getAllAssemblyConfigurations().get(0), mojoParams, new AnsiLogger(new SystemStreamLog(), true, "build"));
         Mockito.verify(assemblyArchiver).createArchive(Mockito.eq(assembly), Mockito.eq("maven"), Mockito.eq("track"), Mockito.any(DockerAssemblyConfigurationSource.class),
             Mockito.eq(false), Mockito.any());
     }
@@ -128,10 +127,10 @@ class DockerAssemblyManagerTest {
 
         BuildImageConfiguration buildConfig = createBuildConfigMultiAssembly();
 
-        AssemblyFiles files = assemblyManager.getAssemblyFiles("testImage", buildConfig.getAssemblyConfigurations().get(0), mojoParams,
+        AssemblyFiles files = assemblyManager.getAssemblyFiles("testImage", buildConfig.getAllAssemblyConfigurations().get(0), mojoParams,
             new AnsiLogger(new SystemStreamLog(), true, "build"));
         Assertions.assertNotNull(files);
-        files = assemblyManager.getAssemblyFiles("testImage", buildConfig.getAssemblyConfigurations().get(1), mojoParams, new AnsiLogger(new SystemStreamLog(), true, "build"));
+        files = assemblyManager.getAssemblyFiles("testImage", buildConfig.getAllAssemblyConfigurations().get(1), mojoParams, new AnsiLogger(new SystemStreamLog(), true, "build"));
         Assertions.assertNotNull(files);
     }
 

--- a/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
@@ -947,16 +947,6 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
         validateRunConfiguration(resolved.getRunConfiguration());
     }
 
-    @Test
-    void testResolveAssembly() {
-
-    }
-
-    @Test
-    void testResolveAssembliesWithSingleAssembly() {
-
-    }
-
     @Override
     protected String getEnvPropertyFile() {
         return "/tmp/envProps.txt";

--- a/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandlerTest.java
@@ -29,6 +29,8 @@ import io.fabric8.maven.docker.config.RunImageConfiguration;
 import io.fabric8.maven.docker.config.UlimitConfig;
 import io.fabric8.maven.docker.config.WaitConfiguration;
 import io.fabric8.maven.docker.config.handler.AbstractConfigHandlerTest;
+import io.fabric8.maven.docker.util.Logger;
+
 import org.apache.maven.plugins.assembly.model.Assembly;
 import org.apache.maven.plugins.assembly.model.DependencySet;
 import org.apache.maven.project.MavenProject;
@@ -658,8 +660,9 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
     void testAssembly() {
         List<ImageConfiguration> configs = resolveImage(imageConfiguration, props(getTestAssemblyData()));
         Assertions.assertEquals(1, configs.size());
+        configs.get(0).initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
 
-        List<AssemblyConfiguration> assemblies = configs.get(0).getBuildConfiguration().getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblies = configs.get(0).getBuildConfiguration().getAllAssemblyConfigurations();
         Assertions.assertEquals(1, assemblies.size());
 
         AssemblyConfiguration config = assemblies.get(0);
@@ -673,8 +676,9 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
     void testMultipleAssemblies() {
         List<ImageConfiguration> configs = resolveImage(imageConfiguration, props(getTestMultipleAssemblyData()));
         Assertions.assertEquals(1, configs.size());
+        configs.get(0).initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
 
-        List<AssemblyConfiguration> assemblies = configs.get(0).getBuildConfiguration().getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblies = configs.get(0).getBuildConfiguration().getAllAssemblyConfigurations();
         Assertions.assertEquals(2, assemblies.size());
 
         AssemblyConfiguration config = assemblies.get(0);
@@ -707,10 +711,13 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
 
         List<ImageConfiguration> configs = resolveImage(imageConfiguration, props(getTestAssemblyData()));
         Assertions.assertEquals(1, configs.size());
+        configs.get(0).initAndValidate(ConfigHelper.NameFormatter.IDENTITY, null);
 
-        AssemblyConfiguration config = configs.get(0).getBuildConfiguration().getAssemblyConfiguration();
-        Assertions.assertNotNull(config.getInline());
-        Assertions.assertEquals(1, config.getInline().getDependencySets().size());
+        List<AssemblyConfiguration> assemblyConfigurations = configs.get(0).getBuildConfiguration().getAllAssemblyConfigurations();
+        Assertions.assertEquals(1, assemblyConfigurations.size());
+        AssemblyConfiguration assemblyConfiguration = assemblyConfigurations.get(0);
+        Assertions.assertNotNull(assemblyConfiguration.getInline());
+        Assertions.assertEquals(1, assemblyConfiguration.getInline().getDependencySets().size());
     }
 
     @Test
@@ -940,6 +947,16 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
         validateRunConfiguration(resolved.getRunConfiguration());
     }
 
+    @Test
+    void testResolveAssembly() {
+
+    }
+
+    @Test
+    void testResolveAssembliesWithSingleAssembly() {
+
+    }
+
     @Override
     protected String getEnvPropertyFile() {
         return "/tmp/envProps.txt";
@@ -1023,7 +1040,7 @@ class PropertyConfigHandlerTest extends AbstractConfigHandlerTest {
          * validate only the descriptor is required and defaults are all used, 'testAssembly' validates
          * all options can be set
          */
-        List<AssemblyConfiguration> assemblyConfigurations = buildConfig.getAssemblyConfigurations();
+        List<AssemblyConfiguration> assemblyConfigurations = buildConfig.getAllAssemblyConfigurations();
         Assertions.assertEquals(1, assemblyConfigurations.size());
 
         AssemblyConfiguration assemblyConfig = assemblyConfigurations.get(0);

--- a/src/test/java/io/fabric8/maven/docker/service/ArchiveServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/ArchiveServiceTest.java
@@ -1,0 +1,56 @@
+package io.fabric8.maven.docker.service;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import org.apache.maven.plugins.assembly.model.Assembly;
+import org.apache.maven.plugins.assembly.model.FileItem;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import io.fabric8.maven.docker.assembly.AssemblyFiles;
+import io.fabric8.maven.docker.assembly.DockerAssemblyManager;
+import io.fabric8.maven.docker.config.AssemblyConfiguration;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.maven.docker.util.MojoParameters;
+
+@ExtendWith(MockitoExtension.class)
+class ArchiveServiceTest {
+
+    @Mock
+    private DockerAssemblyManager dockerAssemblyManager;
+
+    @Mock
+    private Logger log;
+
+    private ArchiveService archiveService;
+
+    @BeforeEach
+    void setUp() {
+        archiveService = new ArchiveService(dockerAssemblyManager, null);
+    }
+
+    @Test
+    void testGetAssemblyFiles() throws Exception {
+        // ARRANGE
+        when(dockerAssemblyManager.getAssemblyFiles(any(), any(), any(), any())).thenReturn(mock(AssemblyFiles.class));
+        AssemblyConfiguration assemblyConfiguration = new AssemblyConfiguration();
+        BuildImageConfiguration build = new BuildImageConfiguration.Builder().assembly(assemblyConfiguration).build();
+        ImageConfiguration imageConfiguration = new ImageConfiguration.Builder().buildConfig(build).build();
+
+        // ACT
+        AssemblyFiles assemblyFiles = archiveService.getAssemblyFiles(imageConfiguration, assemblyConfiguration.getName(), mock(MojoParameters.class));
+
+        // ASSERT
+        Assertions.assertNotNull(assemblyFiles);
+    }
+}

--- a/src/test/java/io/fabric8/maven/docker/service/BuildServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/BuildServiceTest.java
@@ -1,13 +1,13 @@
 package io.fabric8.maven.docker.service;
 
-import io.fabric8.maven.docker.access.BuildOptions;
-import io.fabric8.maven.docker.access.DockerAccess;
-import io.fabric8.maven.docker.access.DockerAccessException;
-import io.fabric8.maven.docker.config.BuildImageConfiguration;
-import io.fabric8.maven.docker.config.CleanupMode;
-import io.fabric8.maven.docker.config.ImageConfiguration;
-import io.fabric8.maven.docker.util.Logger;
-import io.fabric8.maven.docker.util.MojoParameters;
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.project.MavenProject;
 import org.junit.jupiter.api.Assertions;
@@ -20,13 +20,16 @@ import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
 
-import java.io.File;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
-import java.util.Properties;
+import io.fabric8.maven.docker.access.BuildOptions;
+import io.fabric8.maven.docker.access.DockerAccess;
+import io.fabric8.maven.docker.access.DockerAccessException;
+import io.fabric8.maven.docker.assembly.DockerAssemblyManager;
+import io.fabric8.maven.docker.config.AssemblyConfiguration;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.CleanupMode;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.maven.docker.util.MojoParameters;
 
 @ExtendWith(MockitoExtension.class)
 class BuildServiceTest {
@@ -186,6 +189,31 @@ class BuildServiceTest {
         //verify that tries to pull both images
         verifyImagePull(buildConfig, pullManager, buildContext, "fabric8/s2i-java");
         verifyImagePull(buildConfig, pullManager, buildContext, "fabric8/s1i-java");
+    }
+
+    @Test
+    void testBuildImagePullsDefaultImageWhenNoFromImage() throws Exception {
+        BuildImageConfiguration buildConfig = new BuildImageConfiguration.Builder()
+                .build();
+
+        buildConfig.initAndValidate(logger);
+
+        imageConfig = new ImageConfiguration.Builder()
+            .name("test")
+            .buildConfig(buildConfig)
+            .build();
+
+        final ImagePullManager pullManager = new ImagePullManager(null, null, null);
+        final BuildService.BuildContext buildContext = new BuildService.BuildContext.Builder()
+            .mojoParameters(mojoParameters)
+            .build();
+
+        mockMavenProject();
+
+        File buildArchive = buildService.buildArchive(imageConfig, buildContext, "");
+        buildService.buildImage(imageConfig, pullManager, buildContext, buildArchive);
+
+        verifyImagePull(buildConfig, pullManager, buildContext, DockerAssemblyManager.DEFAULT_DATA_BASE_IMAGE);
     }
 
     @Test

--- a/src/test/java/io/fabric8/maven/docker/service/JibBuildServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/JibBuildServiceTest.java
@@ -1,14 +1,16 @@
 package io.fabric8.maven.docker.service;
 
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+
 import com.google.cloud.tools.jib.api.Credential;
-import io.fabric8.maven.docker.access.AuthConfig;
-import io.fabric8.maven.docker.assembly.DockerAssemblyManager;
-import io.fabric8.maven.docker.config.BuildImageConfiguration;
-import io.fabric8.maven.docker.config.ImageConfiguration;
-import io.fabric8.maven.docker.util.AuthConfigFactory;
-import io.fabric8.maven.docker.util.JibServiceUtil;
-import io.fabric8.maven.docker.util.Logger;
-import io.fabric8.maven.docker.util.MojoParameters;
+import com.google.cloud.tools.jib.api.TarImage;
+
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.settings.Settings;
@@ -22,12 +24,14 @@ import org.mockito.MockedStatic;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
 
-import java.io.File;
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.util.Collections;
-import java.util.List;
+import io.fabric8.maven.docker.access.AuthConfig;
+import io.fabric8.maven.docker.assembly.DockerAssemblyManager;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.util.AuthConfigFactory;
+import io.fabric8.maven.docker.util.JibServiceUtil;
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.maven.docker.util.MojoParameters;
 
 @ExtendWith(MockitoExtension.class)
 class JibBuildServiceTest {
@@ -149,6 +153,30 @@ class JibBuildServiceTest {
             List<ImageConfiguration> emptyList = Collections.emptyList();
             // Then
             Assertions.assertDoesNotThrow(() -> jibBuildService.push(emptyList, 1, null, false));
+        }
+    }
+
+    @Test
+    void testBuildCallsBuildContainer(@TempDir Path tmpDir) throws Exception {
+        // ARRANGE
+        Mockito.when(project.getProperties()).thenReturn(new Properties());
+        setupServiceHubExpectations(tmpDir.toFile());
+        setupDockerAssemblyExpectations(tmpDir);
+        final RegistryService.RegistryConfig registryConfig = new RegistryService.RegistryConfig.Builder()
+                .authConfigFactory(authConfigFactory)
+                .build();
+        mockAuthConfigFactory(true, registryConfig);
+
+        JibBuildService jibBuildService = new JibBuildService(serviceHub, params, logger);
+        ImageConfiguration imageConfiguration = getImageConfiguration();
+
+        try (MockedStatic<JibServiceUtil> jibServiceUtilMock = Mockito.mockStatic(JibServiceUtil.class)) {
+            jibServiceUtilMock.when(() -> JibServiceUtil.getBaseImage(imageConfiguration)).thenCallRealMethod();
+            // ACT
+            jibBuildService.build("docker", imageConfiguration, registryConfig);
+
+            // ASSERT
+            jibServiceUtilMock.verify(() -> JibServiceUtil.buildContainer(Mockito.any(), Mockito.any(), Mockito.any()), Mockito.times(1));
         }
     }
 

--- a/src/test/java/io/fabric8/maven/docker/service/WatchServiceTest.java
+++ b/src/test/java/io/fabric8/maven/docker/service/WatchServiceTest.java
@@ -1,0 +1,90 @@
+package io.fabric8.maven.docker.service;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.maven.model.Build;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.logging.SystemStreamLog;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.MockSettings;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import io.fabric8.maven.docker.access.DockerAccess;
+import io.fabric8.maven.docker.access.DockerAccessException;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.config.RunImageConfiguration;
+import io.fabric8.maven.docker.config.WatchImageConfiguration;
+import io.fabric8.maven.docker.config.WatchMode;
+import io.fabric8.maven.docker.model.Image;
+import io.fabric8.maven.docker.util.AnsiLogger;
+import io.fabric8.maven.docker.util.AuthConfigFactory;
+import io.fabric8.maven.docker.util.Logger;
+
+@ExtendWith(MockitoExtension.class)
+class WatchServiceTest {
+    @Mock
+    private ArchiveService archiveService;
+    @Mock
+    private BuildService buildService;
+    @Mock
+    private DockerAccess dockerAccess;
+    @Mock
+    private MojoExecutionService mojoExecutionService;
+    @Mock
+    private QueryService queryService;
+    @Mock
+    private RunService runService;
+    @Mock
+    private Logger log;
+    private WatchService watchService;
+
+
+    @BeforeEach
+    void setUp() {
+        watchService = new WatchService(archiveService, buildService, dockerAccess, mojoExecutionService, queryService, runService, log);
+    }
+
+    @Test
+    void testWatchCallsGetAllAssemblyConfigurations() {
+        // ARRANGE
+        ImageConfiguration imageConfiguration = mock(ImageConfiguration.class);
+        BuildImageConfiguration buildImageConfiguration = mock(BuildImageConfiguration.class);
+        when(imageConfiguration.getBuildConfiguration()).thenReturn(buildImageConfiguration);
+        when(runService.getImagesConfigsInOrder(any(), any())).thenReturn(Collections.singletonList(imageConfiguration));
+
+        WatchService.WatchContext watchContext = mock(WatchService.WatchContext.class);
+        when(watchContext.getWatchMode()).thenReturn(WatchMode.none);
+        BuildService.BuildContext buildContext = mock(BuildService.BuildContext.class);
+        List<ImageConfiguration> imageConfigurations = Collections.emptyList();
+
+        // ACT
+        Executors.newSingleThreadExecutor().submit(() -> {
+            try {
+                watchService.watch(watchContext, buildContext, imageConfigurations);
+            } catch (Exception ignored) {}
+        });
+
+        // ASSERT
+        verify(buildImageConfiguration, timeout(10_000).times(1)).getAllAssemblyConfigurations();
+    }
+
+}

